src/services/cronJobs/fillSnapshotBalances.ts

```diff
@@ -46,88 +49,122 @@ export async function addFillPowerSnapshotBalanceJobsToQueue() {
   let offset = 0;
   let isFinished = false;
 
-  const groupByBlockNumbers: {
-    [p: number]: {
+  const groupByTimestamp: {
+    [timestamp: number]: {
       userId: number;
       powerSnapshotId: number;
       walletAddress: string;
     }[];
   } = {};
+
   while (!isFinished) {
     const powerBoostings = await getPowerBoostingSnapshotWithoutBalance(
-      100,
+      20,
       offset,
     );
-    offset += powerBoostings.length;
-    if (powerBoostings.length === 0) {
-      isFinished = true;
-    }
     powerBoostings.forEach(pb => {
-      if (groupByBlockNumbers[pb.blockNumber]) {
-        groupByBlockNumbers[pb.blockNumber].push(pb);
+      // const timestampInStr = String(Math.floor(pb.time.getTime() / 1000));
+      const timestampInStr = String(pb.time.getTime());
+      if (groupByTimestamp[timestampInStr]) {
+        groupByTimestamp[timestampInStr].push(pb);
       } else {
-        groupByBlockNumbers[pb.blockNumber] = [pb];
+        groupByTimestamp[timestampInStr] = [pb];
       }
     });
-  }
-  logger.debug('jobs', {
-    groupByBlockNumbers: JSON.stringify(groupByBlockNumbers, null, 2),
-  });
-  Object.keys(groupByBlockNumbers).forEach(key => {
-    const chunkSize = 50;
-    const jobDataArray = groupByBlockNumbers[key];
-    for (let i = 0; i < jobDataArray.length; i += chunkSize) {
-      // Our batches length would not be greater than 50, so we convert the array to multiple chunks and add them to queue
-      const chunk = jobDataArray.slice(i, i + chunkSize);
-
-      fillSnapshotBalanceQueue.add({
-        blockNumber: Number(key),
-        data: chunk,
+    // logger.debug('Trying to fill incomplete powerBoostingSnapshots', {
+    //   offset,
+    //   powerBoostingsLength: powerBoostings?.length,
+    // });
+    if (powerBoostings.length === 0) {
+      logger.debug('isFinished true', {
+        offset,
       });
+      isFinished = true;
+      break;
     }
-  });
+    offset += powerBoostings.length;
+
+    // logger.debug('**powerBoostings**', powerBoostings);
+    logger.debug('groupByTimestamp', groupByTimestamp);
+    Object.keys(groupByTimestamp).forEach(key => {
+      const powerSnapshotId = groupByTimestamp[key][0].powerSnapshotId;
+      logger.debug('inside forEach', {
+        timestamp: Number(key),
+        powerSnapshotId,
+      });
+      const jobData = {
+        timestamp: Number(key),
+        // timestamp: Number(Math.floor(Number(key) / 1000)),
+        powerSnapshotId,
+        data: groupByTimestamp[key].map(pb => {
+          return {
+            userId: pb.userId,
+            walletAddress: pb.walletAddress.toLowerCase(),
+          };
+        }),
+      };
+      logger.debug('job data', jobData);
+      fillSnapshotBalanceQueue.add(jobData);
+    });
+  }
```

Comment:
This logic must be outside of the `while` loop. As I see we have reordered some logics in this file which may cause bugs

Reply:
One thing that looks obvious is that the `groupByTimestamp` is never emptied... and in each iteration of loop, all its content will be processed that part of it is repetitive.